### PR TITLE
Sched deck

### PR DESCRIPTION
**Pull request by joakim-hove**
This should be the first PR to be merged in the #2176 effort. With this PR the keywords in the Schedule section are internalized in a new class `ScheduleDeck` - which is essentially a somewhat glorified `std::vector<DeckKeyword>` - this new container is owned by the `Schedule` class and will have application lifetime.

For now the new `ScheduleDeck`will be used to assemble to full `Schedule` datastructure, but in the future the plan is to be able to manipulate this structure directly by adding keywords and "play it again Sam".

### CODE REVIEW

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2999 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2999 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2999 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2999 please

**Comment by joakim-hove** [comment]
jenkins build this  please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Review by bska — COMMENTED** [review]
I freely admit that I don't understand the big picture here.  How are clients, especially the primary simulation loop in opm-simulators, supposed to interact with these structures when it's more developed/closer to its final state?  Do clients need to be aware of `ScheduleBlock`s or are those supposed to be entirely internal to the construction of the higher-level `Schedule` structure?

In what way do you intend to incorporate dynamically triggered actions (e.g., creating new wells or reservoir connections)?  How would WELPI triggered from ACTIONX behave in this new structure?

Other than that, I noticed a number of smaller issues that you can address in the interim.

File: opm/parser/eclipse/EclipseState/Schedule/ScheduleDeck.hpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
+/*
+  Copyright 2021 Statoil ASA.
```
[the remark this answers is not in the document]
Comment (joakim-hove) [reply opm/parser/eclipse/EclipseState/Schedule/ScheduleDeck.hpp:2]: OK

File: src/opm/parser/eclipse/EclipseState/Schedule/ScheduleDeck.cpp [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    m_location(location)
+{
+}
+
+std::size_t ScheduleBlock::size() const {
+    return this->m_keywords.size();
+}
+
+void ScheduleBlock::push_back(const DeckKeyword& keyword) {
+    return this->m_keywords.push_back(keyword);
```
[the remark this answers is not in the document]
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Schedule/ScheduleDeck.cpp:45]: Historical mistake

File: opm/parser/eclipse/EclipseState/Schedule/ScheduleDeck.hpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    public:
+        ScheduleBlock() = default;
+        ScheduleBlock(const KeywordLocation& location, ScheduleTimeType time_type, const std::chrono::system_clock::time_point& start_time);
+        std::size_t size() const;
+        void push_back(const DeckKeyword& keyword);
+        std::optional<DeckKeyword> get(const std::string& kw) const;
+        const std::chrono::system_clock::time_point& start_time() const;
+        const std::chrono::system_clock::time_point& end_time() const;
+        void end_time(const std::chrono::system_clock::time_point& t);
+        ScheduleTimeType time_type() const;
```
[the remark this answers is not in the document]
Comment (joakim-hove) [reply opm/parser/eclipse/EclipseState/Schedule/ScheduleDeck.hpp:42]: Yes.

File: src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp [hunk 4]
```diff
@@ -414,111 +404,82 @@ class ScheduleLogger {
...
+                                                 days,
+                                                 time_unit,
+                                                 start_date,
+                                                 days_total));
+
+                logger(fmt::format("Initializing report step {}/{} at {} ({} {}) - line {}",
+                                   currentStep + 1,
+                                   this->size(),
+                                   start_date,
+                                   convert_time(time_map.getTimePassedUntil(currentStep)),
```
[the remark this answers is not in the document]
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp:426]: You are right.

**Comment by joakim-hove** [comment]
> Other than that, I noticed a number of smaller issues that you can address in the interim.

OK - thank you for the review; I have addressed (nearly ...) all your comments.

**Comment by joakim-hove** [comment]
jenkins build this please

**Review by bska — APPROVED** [review]
Thanks for the updates.  I still don't quite agree that `ScheduleBlock::time_type()` should be `public`, but as these are intended to be internal implementation types it's less of an issue.

I'll merge this into master now.

**Pull request merged**